File: apps/ios/TaakHelden/Core/Sync/ParentSyncCoordinator.swift
import Foundation
import Observation

enum ParentSyncTrigger: String, Equatable {
    case appBecameActive
    case manualRefresh
    case approvalResolved
    case backgroundPush
    case websocketReconnect
    case settingsChanged
}

enum ParentSyncState: Equatable {
    case idle
    case syncing(trigger: ParentSyncTrigger)
    case synced(trigger: ParentSyncTrigger, at: Date)
    case failed(trigger: ParentSyncTrigger, message: String)
}

@Observable
final class ParentSyncCoordinator {
    var state: ParentSyncState = .idle
    var lastTrigger: ParentSyncTrigger?

    func begin(_ trigger: ParentSyncTrigger) {
        lastTrigger = trigger
        state = .syncing(trigger: trigger)
    }

    func finish(_ trigger: ParentSyncTrigger, at date: Date = .now) {
        lastTrigger = trigger
        state = .synced(trigger: trigger, at: date)
    }

    func fail(_ trigger: ParentSyncTrigger, message: String) {
        lastTrigger = trigger
        state = .failed(trigger: trigger, message: message)
    }
}

File: apps/android/core/src/main/kotlin/nl/taakhelden/core/parent/ParentSyncCoordinator.kt
package nl.taakhelden.core.parent

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.time.Instant

public enum class ParentSyncTrigger {
    APP_BECAME_ACTIVE,
    MANUAL_REFRESH,
    APPROVAL_RESOLVED,
    BACKGROUND_PUSH,
    WEBSOCKET_RECONNECT,
    SETTINGS_CHANGED,
}

public sealed interface ParentSyncState {
    public data object Idle : ParentSyncState
    public data class Syncing(val trigger: ParentSyncTrigger) : ParentSyncState
    public data class Synced(val trigger: ParentSyncTrigger, val at: Instant) : ParentSyncState
    public data class Failed(val trigger: ParentSyncTrigger, val message: String) : ParentSyncState
}

/** Tracks *why* the parent dashboard last refreshed, so the header can explain itself. */
public class ParentSyncCoordinator {
    private val _state = MutableStateFlow<ParentSyncState>(ParentSyncState.Idle)
    public val state: StateFlow<ParentSyncState> = _state.asStateFlow()

    private val _lastTrigger = MutableStateFlow<ParentSyncTrigger?>(null)
    public val lastTrigger: StateFlow<ParentSyncTrigger?> = _lastTrigger.asStateFlow()

    public fun begin(trigger: ParentSyncTrigger) {
        _lastTrigger.value = trigger
        _state.value = ParentSyncState.Syncing(trigger)
    }

    public fun finish(trigger: ParentSyncTrigger, at: Instant = Instant.now()) {
        _lastTrigger.value = trigger
        _state.value = ParentSyncState.Synced(trigger, at)
    }

    public fun fail(trigger: ParentSyncTrigger, message: String) {
        _lastTrigger.value = trigger
        _state.value = ParentSyncState.Failed(trigger, message)
    }
}
